discovery-provider/src/challenges/challenge.py

```diff
@@ -0,0 +1,116 @@
+import logging
+from abc import ABC, abstractmethod
+from src.models import Challenge, UserChallenge
+
+logger = logging.getLogger(__name__)
+
+## DB Accessors
+def fetch_user_challenges(session, challenge_id, user_ids):
+    return (session
+            .query(UserChallenge)
+            .filter(
+                UserChallenge.challenge_id == challenge_id,
+                UserChallenge.user_id.in_(user_ids)
+            )
+            ).all()
+
+class ChallengeUpdater(ABC):
+    """`ChallengeUpdater` is an abstract class which provides challenge specific logic 
+    to an instance of a `ChallengeManager`. The only required override is update_user_challenges
+    """
+    @abstractmethod
+    def update_user_challenges(self, session, event, user_challenges, challenge):
+        """This is the main required method to fill out when implementing a new challenge.
+        Given an event type, a list of existing user challenges, and the base challenge type,
+        update the given user_challenges.
+        """
+
+    def on_after_challenge_creation(self, session, user_ids):
+        """Optional method to do some work after the `ChallengeManager` creates new challenges. 
+        If a challenge is backed by it's own table, for instance, create those rows here.
+        """
+
+    def generate_specifier(self, user_id):
```

Comment:
won't this sometimes not have a user_id?

Reply:
right - in that case you override this method. it's always passed a user-id in case we want it